polkadot/xcm/xcm-builder/src/origin_aliases.rs

```diff
@@ -38,3 +38,34 @@ impl<Prefix: Contains<Location>> ContainsPair<Location, Location>
 		false
 	}
 }
+
+/// Alias a descendant location of the original origin.
+pub struct AliasChildLocation;
+impl ContainsPair<Location, Location> for AliasChildLocation {
+	fn contains(origin: &Location, target: &Location) -> bool {
+		return target.starts_with(origin)
+	}
+}
+
+/// Alias a location if it passes `Filter` and the original origin is root of `Origin`.
+///
+/// This can be used to allow (trusted) system chains root to alias into other locations.
+/// **Warning**: do not use with untrusted `Origin` chains.
+pub struct AliasOriginRootUsingFilter<Origin, Filter>(PhantomData<(Origin, Filter)>);
+impl<Origin, Filter> ContainsPair<Location, Location> for AliasOriginRootUsingFilter<Origin, Filter>
+where
+	Origin: Get<Location>,
+	Filter: Contains<Location>,
+{
+	fn contains(origin: &Location, target: &Location) -> bool {
+		// check that `origin` is a root location
+		match origin.unpack() {
+			(1, [Parachain(_)]) |
+			(2, [GlobalConsensus(_)]) |
+			(2, [GlobalConsensus(_), Parachain(_)]) => (),
```

Comment:
Still in the middle of the review and don't understand all the implications, but shouldn't we add also local relay chain `(1, [])` pattern here?

```suggestion
			(1, []) |
			(1, [Parachain(_)]) |
			(2, [GlobalConsensus(_)]) |
			(2, [GlobalConsensus(_), Parachain(_)]) => (),
```

Reply:
The usage pattern for this alias filter is to be used by some parachain to allow a system chain root to proxy the origin of another location. 

Valid/real potential scenarios:

1. some Polkadot parachain to allow Asset Hub root to proxy any other location for a transfer that is going through Asset Hub:
```
// Allow trusted Asset Hub root to alias other locations.
type Aliasers = AliasOriginRootUsingFilter<SystemAssetHubLocation, Everything>;
```

2. Asset Hub to allow Ethereum to proxy for another child location (this can also be achieved with `AliasChildLocation`):
```
// Allow Ethereum root to alias other Ethereum locations.
type Aliasers = AliasOriginRootUsingFilter<EthereumLocation, StartsWith<EthereumLocation>>;
// or
type Aliasers = AliasChildLocation;
```

3. Asset Hub Polkadot to allow Asset Hub Kusama to proxy another Kusama origin for a transfer that has to go through Asset Hub Kusama:
```
// Allow KAH root to alias other Kusama locations.
type Aliasers = AliasOriginRootUsingFilter<KusamaAssetHubLocation, StartsWith<GlobalKusama>>;
``` 

Transfers should not be routed through Relay Chain in practice, so there is no real use case for having Relay root to proxy any other location.

Reply:
ok, thank you, got it